scipy/linalg/special_matrices.py

```diff
@@ -1190,3 +1190,121 @@ def fiedler_companion(a):
     c[[0, 1], 0] = [-a[1], 1]
 
     return c
+
+
+def convolution_matrix(a, n, mode='full'):
+    """
+    Construct a convolution matrix.
+
+    Constructs the dense matrix representing one-dimensional
+    convolution [1]_.
+
+    ``A = convolution_matrix(a, n[, mode])`` creates a matrix ``A`` such
+    that ``A @ v`` (or ``matmul(A, v)``) is equivalent to using
+    ``convolve(a, v[, mode])``. In the default 'full' mode, the entries of
+    ``A`` are given by ``A[i, j] == (a[i-j] if (0 <= (i-j) < m) else 0)``.
```

Comment:
I've read this part more than 5 times and still can't picture anything concrete. Same applies to np.convolve docstring too. Could we use more verbose description something along the lines of? 

```python
def convolution_matrix(a, n, mode='full'):
    """
    Construct a convolution matrix.

    Constructs the dense matrix representing one-dimensional convolution [1]_.
    ``A = convolution_matrix(a, n[, mode])`` creates a matrix ``A`` such that
    ``A @ v`` is equivalent to using ``convolve(a, v[, mode])``. 

    Suppose the input array is a triplet of ``[x, y, z]``. In the default
    'full' mode, partial overlaps (or edge effects) are allowed

        [x, 0, 0, ..., 0, 0]
        [y, x, 0, ..., 0, 0]
        [z, y, x, ..., 0, 0]
        ...

    In 'valid' mode, only the rows where complete overlap occurs, are returned,
    hence the top two rows in the example above are invalid. In 'same' mode,
    ``max(len(a), n)`` defines the row number and partial overlaps are allowed
     until sufficient rows are generated.

    ....

    """
```

Reply:
Good idea.